File: Solutions/228_Summary_Ranges/228_Summary_Ranges.cs
public class Solution {
    public IList<string> SummaryRanges(int[] nums) {
        int i = 0;
        var result = new List<string>();
        int n = nums.Length;
        
        
        while(i < n) {
            int s = i, e = i;
            while(e < n-1 && nums[e] + 1 == nums[e+1])
                e++;
            if(nums[s] == nums[e])
                result.Add(new String($"{nums[e]}"));
            else
                result.Add(new String($"{nums[s]}->{nums[e]}"));
            i = e + 1;
        }
        
        return result;
    }
}

//Third fastest
public class Solution {
    public IList<string> SummaryRanges(int[] nums) {
        var ans = new List<string>();
        
        if(nums.Length == 0)
        {
            return ans;
        }
        
        int start = nums[0];
        int prev = nums[0];
        for(int i = 1; i<nums.Length ; i++)
        {
            if(nums[i] == prev + 1)
            {
                prev = nums[i];
                continue;
            }
            ans.Add(GenerateRange(start, prev));    
            start = nums[i];
            prev = nums[i];
        }
        ans.Add(GenerateRange(start, prev));
        return ans;
    }
    
    private string GenerateRange(int start, int end)
    {
        if(start == end){
            return start.ToString();
        }
        
        return String.Format("{0}->{1}", start, end);
    }
}

//Second fastest
public class Solution {
    public IList<string> SummaryRanges(int[] nums) {
        IList<string> res=new List<string>();
        if(nums==null||nums.Length==0) return res;
        int slow=0,fast=0;
        while(fast<nums.Length)
        {
            if(fast+1<nums.Length&&nums[fast]+1==nums[fast+1]) fast++;
            else if(slow==fast){res.Add(nums[slow].ToString());slow=fast+1;fast=fast+1;}
            else {res.Add(nums[slow].ToString()+"->"+nums[fast].ToString());slow=fast+1;fast=fast+1;}
        }
        return res;        
    }
}

//Fastest
public class Solution {
    public IList<string> SummaryRanges(int[] nums) {
        
        IList<string> summary = new List<string>();
        
        for (int i = 0, j = 0; j < nums.Length; j++) {            
            // check if j + 1 extends the range [nums[i], nums[j]]
            if (j + 1 < nums.Length && nums[j] + 1 == nums[j+1])
                continue;
            
            // put the range [nums[i], nums[j]] into the list
            if (i == j)
                summary.Add(nums[i].ToString());
            else
                summary.Add(nums[i] + "->" + nums[j]);
            
            i = j + 1;
        }
        return summary;        
    }
}

File: Solutions/228_Summary_Ranges/228_Summary_Ranges.py
#Fourth fastest
class Solution:
    def summaryRanges(self, nums: List[int]) -> List[str]:
        ans = []
        
        if not nums:
            return ans
        
        count = 1
        
        for i in range(1, len(nums)):
            if nums[i - 1] + 1 == nums[i]:
                count += 1
            else:
                ans.append((str(nums[i - 1] - count + 1) + "->" if count > 1 else "" ) + str(nums[i - 1]))
                count = 1
                
        ans.append((str(nums[-1] - count + 1) + "->" if count > 1 else "" ) + str(nums[-1]))
        return ans

#Third fastest
class Solution:
    def summaryRanges(self, nums: List[int]) -> List[str]:

#         #Solution#1
#         ans = []
        
#         if (len(nums) == 0): return ans
        
#         start, end = 0, 0
#         for end in range(1, len(nums)):
#             # print("%d~%d" % (start, end))
#             if (nums[start] + (end-start) != nums[end]):
#                 if (start==end-1):
#                     ans.append("%d" % nums[start])
#                 else:
#                     ans.append("%d->%d" % (nums[start], nums[end-1]))
#                 start = end
        
#         # print("%d~%d" % (start, end))
#         if (start==end):
#             ans.append("%d" % nums[start])
#         else:
#             ans.append("%d->%d" % (nums[start], nums[end]))
            
#         return ans
    
        #Solution#1v2
        ans = []
        l = len(nums)
        if (l == 0): return ans
        start = 0
        for end in range(l+1):
            #print("%d~%d" % (start, end))
            if (end == l or 
                nums[start] + (end-start) != nums[end]):
                if (start==end-1): 
                    ans.append("%d" % nums[start])
                else: 
                    ans.append("%d->%d" % (nums[start], nums[end-1]))
                start = end
        return ans

#Second fastest
class Solution:
    def summaryRanges(self, nums: List[int]) -> List[str]:
        ans = []
        if len(nums) == 0:
            return ans
        if len(nums) == 1:
            return [str(nums[0])]
        
        start = nums[0]
        for i in range(1, len(nums)):
            if nums[i] != nums[i - 1] + 1:
                if start == nums[i - 1]:
                    ans.append(str(nums[i - 1]))
                else:
                    ans.append(str(start) + '->' + str(nums[i - 1]))
                start = nums[i]
            if i == len(nums) - 1:
                if start == nums[i]:
                        ans.append(str(start))
                else:
                    ans.append(str(start) + '->' + str(nums[i]))                
            
        return ans

#Fastest
def makeRange(ci, cj):
    if ci == cj:
        return str(ci)
    else:
        return f"{ci}->{cj}"

class Solution:
    
    def summaryRanges(self, nums: List[int]) -> List[str]:
        if not nums:
            return []
        
        idx, ci, cj = nums[0]-1, nums[0], nums[0]
        res = []
        
        for n in nums:
            if n != (idx+1):
                res.append(makeRange(ci, cj))
                ci = idx = n
            else:
                idx += 1
            cj = n
        
        res.append(makeRange(ci, cj))
        return res